# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1460391379 14400
#      Mon Apr 11 12:16:19 2016 -0400
# Node ID 37f6ed8a6f1bfb2dca22e9e482d8072ff37c1982
# Parent  ec01be3b8f5d089234828455dc1bfc26fe37a3f2
store liboctave and liboctinterp hg ids separately

* liboctinterp-build-info.h: Rename from build-info.h.
* liboctinterp-build-info.in.cc: Rename from
build-info.in.cc.
* liboctinterp-build-info.h, liboctinterp-build-info.in.cc
(liboctinterp_hg_id): Rename function from octave_hg_id.
* libinterp/module.mk: Update.

* liboctave/liboctave-build-info.h,
liboctave/liboctave-build-info.in.cc: New files.
* liboctave/module.mk: Update.

* toplev.cc: Include both liboctave-build-info.h and
liboctinterp-build-info.h.
(F__octave_config_info__): Replace hg_id field with liboctave_hg_id
and liboctinterp_hg_id fields.

diff --git a/libinterp/corefcn/toplev.cc b/libinterp/corefcn/toplev.cc
--- a/libinterp/corefcn/toplev.cc
+++ b/libinterp/corefcn/toplev.cc
@@ -47,17 +47,18 @@ along with Octave; see the file COPYING.
 #include "oct-env.h"
 #include "oct-fftw.h"
 #include "oct-locbuf.h"
 #include "quit.h"
 #include "singleton-cleanup.h"
 #include "str-vec.h"
 
 #include "build-env.h"
-#include "build-info.h"
+#include "liboctave-build-info.h"
+#include "liboctinterp-build-info.h"
 #include "defaults.h"
 #include "defun.h"
 #include "error.h"
 #include "file-io.h"
 #include "graphics.h"
 #include "input.h"
 #include "lex.h"
 #include "load-save.h"
@@ -1426,23 +1427,28 @@ specified option.\n\
       { "bindir", subst_octave_home (OCTAVE_BINDIR) },
       { "canonical_host_type", OCTAVE_CANONICAL_HOST_TYPE },
       { "datadir", subst_octave_home (OCTAVE_DATADIR) },
       { "datarootdir", subst_octave_home (OCTAVE_DATAROOTDIR) },
       { "exec_prefix", subst_octave_home (OCTAVE_EXEC_PREFIX) },
       { "fcnfiledir", subst_octave_home (OCTAVE_FCNFILEDIR) },
       { "fftw_version", octave_fftw_version () },
       { "fftwf_version", octave_fftwf_version () },
-      { "hg_id", octave_hg_id () },
       { "imagedir", subst_octave_home (OCTAVE_IMAGEDIR) },
       { "includedir", subst_octave_home (OCTAVE_INCLUDEDIR) },
       { "infodir", subst_octave_home (OCTAVE_INFODIR) },
       { "infofile", subst_octave_home (OCTAVE_INFOFILE) },
       { "libdir", subst_octave_home (OCTAVE_LIBDIR) },
       { "libexecdir", subst_octave_home (OCTAVE_LIBEXECDIR) },
+      // The liboctave and liboctinterp hg ids should always be the
+      // same, but it is possible for someone to accidentally have a
+      // mismatched pair so we record both separately so that we can
+      // ensure that they are consistent.
+      { "liboctave_hg_id", liboctave_hg_id () },
+      { "liboctinterp_hg_id", liboctinterp_hg_id () },
       { "localapiarchlibdir", subst_octave_home (OCTAVE_LOCALAPIARCHLIBDIR) },
       { "localapifcnfiledir", subst_octave_home (OCTAVE_LOCALAPIFCNFILEDIR) },
       { "localapioctfiledir", subst_octave_home (OCTAVE_LOCALAPIOCTFILEDIR) },
       { "localarchlibdir", subst_octave_home (OCTAVE_LOCALARCHLIBDIR) },
       { "localfcnfiledir", subst_octave_home (OCTAVE_LOCALFCNFILEDIR) },
       { "localoctfiledir", subst_octave_home (OCTAVE_LOCALOCTFILEDIR) },
       { "localstartupfiledir", subst_octave_home (OCTAVE_LOCALSTARTUPFILEDIR) },
       { "localverarchlibdir", subst_octave_home (OCTAVE_LOCALVERARCHLIBDIR) },
diff --git a/libinterp/build-info.h b/libinterp/liboctinterp-build-info.h
rename from libinterp/build-info.h
rename to libinterp/liboctinterp-build-info.h
--- a/libinterp/build-info.h
+++ b/libinterp/liboctinterp-build-info.h
@@ -15,20 +15,20 @@ FITNESS FOR A PARTICULAR PURPOSE.  See t
 for more details.
 
 You should have received a copy of the GNU General Public License
 along with Octave; see the file COPYING.  If not, see
 <http://www.gnu.org/licenses/>.
 
 */
 
-#if ! defined (octave_build_info_h)
-#define octave_build_info_h 1
+#if ! defined (octave_liboctinterp_build_info_h)
+#define octave_liboctinterp_build_info_h 1
 
 #include "octave-config.h"
 
 #include <ctime>
 
 #include <string>
 
-extern OCTAVE_API std::string octave_hg_id (void);
+extern OCTAVE_API std::string liboctinterp_hg_id (void);
 
 #endif
diff --git a/libinterp/build-info.in.cc b/libinterp/liboctinterp-build-info.in.cc
rename from libinterp/build-info.in.cc
rename to libinterp/liboctinterp-build-info.in.cc
--- a/libinterp/build-info.in.cc
+++ b/libinterp/liboctinterp-build-info.in.cc
@@ -20,15 +20,15 @@ along with Octave; see the file COPYING.
 <http://www.gnu.org/licenses/>.
 
 */
 
 #ifdef HAVE_CONFIG_H
 #  include "config.h"
 #endif
 
-#include "build-info.h"
+#include "liboctinterp-build-info.h"
 
 std::string
-octave_hg_id (void)
+liboctinterp_hg_id (void)
 {
   return "%OCTAVE_HG_ID%";
 }
diff --git a/libinterp/module.mk b/libinterp/module.mk
--- a/libinterp/module.mk
+++ b/libinterp/module.mk
@@ -38,19 +38,19 @@ GENERATED_MAKE_BUILTINS_INCS = \
   libinterp/corefcn/mxarray.h \
   libinterp/parse-tree/oct-gperf.h \
   libinterp/parse-tree/oct-parse.h \
   libinterp/version.h
 
 BUILT_SOURCES += \
   $(GENERATED_MAKE_BUILTINS_INCS) \
   libinterp/build-env.cc \
-  libinterp/build-info.cc \
   libinterp/builtin-defun-decls.h \
   libinterp/builtins.cc \
+  libinterp/liboctinterp-build-info.cc \
   libinterp/corefcn/oct-errno.cc \
   libinterp/corefcn/oct-tex-lexer.cc \
   libinterp/corefcn/oct-tex-parser.cc \
   libinterp/corefcn/oct-tex-symbols.cc \
   libinterp/operators/ops.cc \
   libinterp/parse-tree/lex.cc \
   libinterp/parse-tree/oct-parse.cc
 
@@ -72,50 +72,50 @@ LIBINTERP_BUILT_DISTFILES = \
   libinterp/parse-tree/oct-gperf.h \
   libinterp/parse-tree/oct-parse.h
 
 ## Files that are created during build process and installed,
 ## BUT not distributed in tarball.
 LIBINTERP_BUILT_NODISTFILES = \
   libinterp/build-env.cc \
   libinterp/build-env-features.cc \
-  libinterp/build-info.cc \
+  libinterp/liboctinterp-build-info.cc \
   libinterp/corefcn/mxarray.h \
   libinterp/corefcn/oct-errno.cc \
   libinterp/corefcn/defaults.h \
   libinterp/corefcn/graphics.h \
   libinterp/builtin-defun-decls.h \
   libinterp/operators/ops.cc \
   libinterp/version.h \
   $(OPT_HANDLERS) \
   $(ALL_DEF_FILES) \
   libinterp/builtins.cc
 
 libinterp_EXTRA_DIST += \
   libinterp/DOCSTRINGS \
   libinterp/build-env.in.cc \
   libinterp/build-env-features.sh \
-  libinterp/build-info.in.cc \
   libinterp/find-defun-files.sh \
   libinterp/gendoc.pl \
   libinterp/genprops.awk \
+  libinterp/liboctinterp-build-info.in.cc \
   libinterp/mk-errno-list \
   libinterp/mk-pkg-add \
   libinterp/mkbuiltins \
   libinterp/mkdefs \
   libinterp/mkops \
   libinterp/version.in.h \
   $(LIBINTERP_BUILT_DISTFILES)
 
 octinclude_HEADERS += \
   libinterp/corefcn/graphics-props.cc \
   libinterp/builtins.h \
   libinterp/builtin-defun-decls.h \
   libinterp/build-env.h \
-  libinterp/build-info.h \
+  libinterp/liboctinterp-build-info.h \
   libinterp/octave.h \
   libinterp/options-usage.h \
   $(OCTAVE_VALUE_INC) \
   $(PARSE_TREE_INC) \
   $(PARSER_INC) \
   $(COREFCN_INC)
 
 noinst_HEADERS += \
@@ -163,17 +163,17 @@ nodist_libinterp_liboctinterp_la_SOURCES
   libinterp/corefcn/oct-errno.cc \
   libinterp/corefcn/defaults.h \
   libinterp/corefcn/graphics.h \
   libinterp/operators/ops.cc \
   libinterp/builtin-defun-decls.h \
   libinterp/builtins.cc \
   libinterp/build-env.cc \
   libinterp/build-env-features.cc \
-  libinterp/build-info.cc \
+  libinterp/liboctinterp-build-info.cc \
   libinterp/version.h
 
 libinterp_liboctinterp_la_LIBADD = \
   libinterp/octave-value/liboctave-value.la \
   libinterp/parse-tree/libparse-tree.la \
   libinterp/corefcn/libcorefcn.la \
   liboctave/liboctave.la \
   $(LIBOCTINTERP_LINK_DEPS)
@@ -258,17 +258,17 @@ libinterp/build-env.cc: libinterp/build-
 libinterp/build-env-features.cc: config.h libinterp/build-env-features.sh | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(SHELL) $(srcdir)/libinterp/build-env-features.sh $< > $@-t && \
 	$(simple_move_if_change_rule)
 
 libinterp/version.h: libinterp/version.in.h build-aux/mk-version-h.sh | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)$(call simple-filter-rule,build-aux/mk-version-h.sh)
 
-libinterp/build-info.cc: libinterp/build-info.in.cc HG-ID | libinterp/$(octave-dirstamp)
+libinterp/liboctinterp-build-info.cc: libinterp/liboctinterp-build-info.in.cc HG-ID | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
 	$(SED) \
 	  -e "s|%NO_EDIT_WARNING%|DO NOT EDIT!  Generated automatically by Makefile|" \
 	  -e "s|%OCTAVE_HG_ID%|`cat $(builddir)/HG-ID`|" $< > $@-t && \
 	$(simple_move_if_change_rule)
 
 libinterp/builtins.cc: $(DEF_FILES) libinterp/mkbuiltins | libinterp/$(octave-dirstamp)
 	$(AM_V_GEN)rm -f $@-t && \
diff --git a/liboctave/liboctave-build-info.h b/liboctave/liboctave-build-info.h
new file mode 100644
--- /dev/null
+++ b/liboctave/liboctave-build-info.h
@@ -0,0 +1,34 @@
+/*
+
+Copyright (C) 2016 M. Muetzel
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+#if ! defined (octave_liboctave_build_info_h)
+#define octave_liboctave_build_info_h 1
+
+#include "octave-config.h"
+
+#include <ctime>
+
+#include <string>
+
+extern OCTAVE_API std::string liboctave_hg_id (void);
+
+#endif
diff --git a/liboctave/liboctave-build-info.in.cc b/liboctave/liboctave-build-info.in.cc
new file mode 100644
--- /dev/null
+++ b/liboctave/liboctave-build-info.in.cc
@@ -0,0 +1,34 @@
+// %NO_EDIT_WARNING%
+/*
+
+Copyright (C) 2016 M. Muetzel
+
+This file is part of Octave.
+
+Octave is free software; you can redistribute it and/or modify it
+under the terms of the GNU General Public License as published by the
+Free Software Foundation; either version 3 of the License, or (at your
+option) any later version.
+
+Octave is distributed in the hope that it will be useful, but WITHOUT
+ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+for more details.
+
+You should have received a copy of the GNU General Public License
+along with Octave; see the file COPYING.  If not, see
+<http://www.gnu.org/licenses/>.
+
+*/
+
+#ifdef HAVE_CONFIG_H
+#  include "config.h"
+#endif
+
+#include "liboctave-build-info.h"
+
+std::string
+liboctave_hg_id (void)
+{
+  return "%OCTAVE_HG_ID%";
+}
diff --git a/liboctave/module.mk b/liboctave/module.mk
--- a/liboctave/module.mk
+++ b/liboctave/module.mk
@@ -1,18 +1,20 @@
-liboctave_EXTRA_DIST =
+liboctave_EXTRA_DIST = \
+  liboctave/liboctave-build-info.in.cc
 
 liboctave_CLEANFILES =
 liboctave_DISTCLEANFILES =
 liboctave_MAINTAINERCLEANFILES =
 
 ## Search local directories before those specified by the user.
 liboctave_liboctave_la_CPPFLAGS = \
   @OCTAVE_DLL_DEFS@ \
   @CRUFT_DLL_DEFS@ \
+  -Iliboctave -I$(srcdir)/liboctave \
   -I$(srcdir)/liboctave/array \
   -I$(srcdir)/liboctave/cruft/misc \
   -Iliboctave/numeric -I$(srcdir)/liboctave/numeric \
   -Iliboctave/operators -I$(srcdir)/liboctave/operators \
   -I$(srcdir)/liboctave/system \
   -I$(srcdir)/liboctave/util \
   -Ilibgnu -I$(srcdir)/libgnu
 
@@ -23,19 +25,24 @@ liboctave_liboctave_la_CXXFLAGS = $(AM_C
 octlib_LTLIBRARIES += liboctave/liboctave.la
 
 BUILT_INCS = \
   $(BUILT_LIBOCTAVE_OPERATORS_INC) \
   $(LIBOCTAVE_OPT_INC)
 
 BUILT_SOURCES += \
   $(BUILT_INCS) \
-  $(BUILT_LIBOCTAVE_OPERATORS_SOURCES)
+  $(BUILT_LIBOCTAVE_OPERATORS_SOURCES) \
+  liboctave/liboctave-build-info.cc
+
+LIBOCTAVE_BUILT_NODISTFILES = \
+  liboctave/liboctave-build-info.cc
 
 octinclude_HEADERS += \
+  liboctave/liboctave-build-info.h \
   $(ARRAY_INC) \
   $(CRUFT_INC) \
   $(NUMERIC_INC) \
   $(LIBOCTAVE_OPERATORS_INC) \
   $(SYSTEM_INC) \
   $(UTIL_INC) \
   $(OTHER_INC) \
   $(LIBOCTAVE_TEMPLATE_SRC)
@@ -56,17 +63,18 @@ include liboctave/array/module.mk
 include liboctave/cruft/module.mk
 include liboctave/numeric/module.mk
 include liboctave/operators/module.mk
 include liboctave/system/module.mk
 include liboctave/util/module.mk
 
 ## liboctave merely collects a bunch of compiled convenience libraries.
 ## It has no source code itself.
-liboctave_liboctave_la_SOURCES =
+liboctave_liboctave_la_SOURCES = \
+  liboctave/liboctave-build-info.cc
 
 # Dummy C++ source to force C++ linking.
 EXTRA_liboctave_liboctave_la_SOURCES = liboctave/.dummy_force_cxx_link.cc
 
 liboctave_liboctave_la_LIBADD += \
   libgnu/libgnu.la \
   $(LIBOCTAVE_LINK_DEPS)
 
@@ -97,22 +105,32 @@ LIBOCTAVE_TST_SRC = \
 LIBOCTAVE_TST_FILES_SRC := $(shell $(SHELL) $(srcdir)/build-aux/find-files-with-tests.sh "$(srcdir)" $(LIBOCTAVE_TST_SRC))
 
 LIBOCTAVE_TST_FILES := $(addsuffix -tst,$(LIBOCTAVE_TST_FILES_SRC))
 
 liboctavetestsdir := $(octtestsdir)
 
 nobase_liboctavetests_DATA = $(LIBOCTAVE_TST_FILES)
 
+liboctave/liboctave-build-info.cc: liboctave/liboctave-build-info.in.cc HG-ID | liboctave/$(octave-dirstamp)
+	$(AM_V_GEN)rm -f $@-t && \
+	$(SED) \
+	  -e "s|%NO_EDIT_WARNING%|DO NOT EDIT!  Generated automatically by Makefile|" \
+	  -e "s|%OCTAVE_HG_ID%|`cat $(builddir)/HG-ID`|" $< > $@-t && \
+	$(simple_move_if_change_rule)
+
 EXTRA_DIST += $(liboctave_EXTRA_DIST)
 
 liboctave_DISTCLEANFILES += \
+  $(LIBOCTAVE_BUILT_NODISTFILES) \
   $(BUILT_INCS) \
   $(LIBOCTAVE_TST_FILES)
 
+BUILT_NODISTFILES += $(LIBOCTAVE_BUILT_NODISTFILES)
+
 CLEANFILES += $(liboctave_CLEANFILES)
 DISTCLEANFILES += $(liboctave_DISTCLEANFILES)
 MAINTAINERCLEANFILES += $(liboctave_MAINTAINERCLEANFILES)
 
 liboctave-clean:
 	rm -f $(liboctave_CLEANFILES)
 
 liboctave-distclean: liboctave-clean
